### PR TITLE
Probabilities calculation method has been aligned with sklearn implementation

### PR DESCRIPTION
**Pull request by masdevas**
# Description
Pull request includes changes in method of probabilities calculation from trained model.

Changes proposed in this pull request:
1) Save samples for each class count on training stage (instead of probabilities distribution saving)
2) Calculate probabilities distribution on predict stage from trained model

### CODE REVIEW

**Comment by masdevas** [comment]
make test

**Comment by ShvetsKS** [comment]
please fix failed tests at first

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 1]
```diff
@@ -729,6 +749,19 @@ Status PredictClassificationTask<algorithmFPType, cpu>::predictByBlocksOfTrees(s
...
         s = safeStat.detach();
     }
+    if(probBDPtr != nullptr)
+    {
+        daal::threader_for(dim.nDataBlocks, dim.nDataBlocks, [&](size_t iBlock)
+        {
+            const size_t iStartRow = iBlock * dim.nRowsInBlock;
+            const size_t nRowsToProcess = (iBlock == dim.nDataBlocks - 1) ? dim.nRowsTotal - iStartRow : dim.nRowsInBlock;
+            algorithmFPType* prob = probBDPtr + iStartRow * _nClasses;
+            daal::threader_for(nRowsToProcess, nRowsToProcess, [&](size_t iRow)
```
Comment (SmirnovEgorRu) [inline algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: You create nRowsToProcess TBB tasks here, it is unefficient way.
Better to create some blocks of rows

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 2]
```diff
@@ -107,8 +107,30 @@ protected:
...
+    size_t getMaxClass(const ClassIndexType* counts) const { return services::internal::getMaxElementIndex<ClassIndexType, cpu>(counts, _nClasses); }
+    void normalizeProbabilities(algorithmFPType* probs) {
+        algorithmFPType sum = 0.0;
+        for(size_t index = 0; index < _nClasses; index++)
+        {
+            sum += probs[index];
+        }
+        sum = 1 / sum;
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
```
Comment (SmirnovEgorRu) [inline algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: These pragmas can even slowdown your code for small dimensions like nClasses

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 3]
```diff
@@ -590,13 +615,8 @@ Status PredictClassificationTask<algorithmFPType, cpu>::predictAllPointsByAllTre
...
             for (size_t iRes = 0; iRes < nRowsToProcess; ++iRes)
             {
-                PRAGMA_IVDEP
-                PRAGMA_VECTOR_ALWAYS
-                for (size_t j = 0; j < _nClasses; j++)
-                {
-                    prob_internal[iRes * _nClasses + j] = algorithmFPType(prob_internal[iRes * _nClasses + j]) * inverseNTreesTotal;
-                }
-                if (_res)
+                normalizeProbabilities(prob_internal + iRes * _nClasses);
```
Comment (SmirnovEgorRu) [inline algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: I think you can merge normalizeProbabilities and getMaxClass at once to speedup computations.

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 4]
```diff
@@ -107,8 +107,30 @@ protected:
...
-    size_t getMaxClass(const ClassIndexType * counts) const { return services::internal::getMaxElementIndex<ClassIndexType, cpu>(counts, _nClasses); }
-
+    size_t getMaxClass(const ClassIndexType* counts) const { return services::internal::getMaxElementIndex<ClassIndexType, cpu>(counts, _nClasses); }
+    void normalizeProbabilities(algorithmFPType* probs) {
+        algorithmFPType sum = 0.0;
+        for(size_t index = 0; index < _nClasses; index++)
+        {
+            sum += probs[index];
+        }
+        sum = 1 / sum;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: It isn't a sum any more.

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 5]
```diff
@@ -107,8 +107,30 @@ protected:
...
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for(size_t index = 0; index < _nClasses; index++)
+        {
+            probs[index] *= sum;
+        }
+    }
+    void normalizeProbabilities(algorithmFPType* probs, size_t countSamplesInLeafs) {
+        algorithmFPType sum = 1.0 / countSamplesInLeafs;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: const? It isn't sum.

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 6]
```diff
@@ -107,8 +107,30 @@ protected:
...
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for(size_t index = 0; index < _nClasses; index++)
```
Comment (averbukh) [inline algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: Fix coding style

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 7]
```diff
@@ -455,10 +479,12 @@ Status PredictClassificationTask<algorithmFPType, cpu>::predictByAllTrees(size_t
             const size_t nRowsToProcess = (iBlock == dim.nDataBlocks - 1) ? dim.nRowsTotal - iStartRow : dim.nRowsInBlock;
             ReadRows<algorithmFPType, cpu> xBD(const_cast<NumericTable *>(_data), iStartRow, nRowsToProcess);
             DAAL_CHECK_BLOCK_STATUS_THR(xBD);
-            algorithmFPType * res  = resBD.get() + iStartRow;
-            algorithmFPType * prob = probPtr + iStartRow * _nClasses;
+            algorithmFPType* res = resBD.get() + iStartRow;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: algorithmFPType * const ?

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 8]
```diff
@@ -455,10 +479,12 @@ Status PredictClassificationTask<algorithmFPType, cpu>::predictByAllTrees(size_t
...
+            algorithmFPType* prob = probPtr + iStartRow * _nClasses;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: algorithmFPType * const ?

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i (hunk 3 above)
Comment (averbukh) [reply algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: +1

File: algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 9]
```diff
@@ -729,6 +749,19 @@ Status PredictClassificationTask<algorithmFPType, cpu>::predictByBlocksOfTrees(s
         });
         s = safeStat.detach();
     }
+    if(probBDPtr != nullptr)
+    {
+        daal::threader_for(dim.nDataBlocks, dim.nDataBlocks, [&](size_t iBlock)
+        {
+            const size_t iStartRow = iBlock * dim.nRowsInBlock;
+            const size_t nRowsToProcess = (iBlock == dim.nDataBlocks - 1) ? dim.nRowsTotal - iStartRow : dim.nRowsInBlock;
+            algorithmFPType* prob = probBDPtr + iStartRow * _nClasses;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: algorithmFPType * const ?

**Comment by ShvetsKS** [comment]
It was wrong assumption that the difference in probability was due to sklearn computes prediction by proposed method (weighted by number of samples of specific class).
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/_forest.py#L667

From link above we can see that each estimator returns values in interval (0,1) already.
The true reason is in training stage: DAAL trees predict similar(exactly the same) class label (3) for "unknown" sample while sklearn trees in ensemble predict different class labels(2,3,4).
